src/mlpack/tests/ann_layer_test.cpp

```diff
@@ -779,6 +780,92 @@ BOOST_AUTO_TEST_CASE(GradientLSTMLayerTest)
 }
 
 /**
+ * GRU layer numerically gradient test.
+ */
+BOOST_AUTO_TEST_CASE(GradientGRULayerTest)
+{
+  // GRU function gradient instantiation.
+  struct GradientFunction
+  {
+    GradientFunction()
+    {
+      input = arma::randu(5, 1);
+      target = arma::mat("1; 1; 1; 1; 1");
+      const size_t rho = 5;
+
+      model = new RNN<NegativeLogLikelihood<> >(input, target, rho);
+      model->Add<IdentityLayer<> >();
+      model->Add<Linear<> >(1, 10);
+      model->Add<GRU<> >(10, 3, rho);
+      model->Add<LogSoftMax<> >();
+    }
+
+    ~GradientFunction()
+    {
+      delete model;
+    }
+
+    double Gradient(arma::mat& gradient) const
+    {
+      arma::mat output;
+      double error = model->Evaluate(model->Parameters(), 0);
+      model->Gradient(model->Parameters(), 0, gradient);
+      return error;
+    }
+
+    arma::mat& Parameters() { return model->Parameters(); }
+
+    RNN<NegativeLogLikelihood<> >* model;
+    arma::mat input, target;
+  } function;
+
+  BOOST_REQUIRE_LE(CheckGradient(function), 1e-4);
+}
+
+/**
+ * GRU layer manual forward test.
+ */
+BOOST_AUTO_TEST_CASE(ForwardGRULayerTest)
+{
+  GRU<> gru(3, 3, 5);
+
+  NetworkInitialization<ConstInitialization<1>> networkInit;
+  networkInit.Initialize(gru.Model(), gru.Parameters());
+
+  arma::mat input = arma::ones(3, 1);
+  arma::mat output;
+  gru.Forward(std::move(input), std::move(output));
+
+  arma::mat expectedOutput = arma::ones(3, 1);
+  expectedOutput *= -4;
+  expectedOutput = arma::exp(expectedOutput);
+  expectedOutput = arma::ones(3, 1) / (arma::ones(3, 1) + expectedOutput);
+  expectedOutput = (arma::ones(3, 1)  - expectedOutput) % expectedOutput;
```

Comment:
For Backward() test, the code is definitely non-trivial but the way we test gradients, if the Backward function is not correct for all the layers, the gradients of the input won't be correct. The gradients computed at each layer depends on the signal coming from the layer above, which is computed by the backward call. Unless the backward function of the layer above is correct, the gradient of the layer below won't be correct. Writing the test is not a very complex task, but I feel that it is unnecessary.